aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp

```diff
@@ -640,31 +718,71 @@ void svd_cusolver(const Tensor& A,
   const auto n = A.size(-1);
   const auto k = std::min(m, n);
 
-  // heuristic for using `gesvdjBatched` over `gesvdj`
-  // gesvdjBatched just works for square matrices
-  // Need to pass a copy of A, since A will be rewritten inside the function call
-  if (m <= 32 && n <= 32 && batch_size > 1 && (full_matrices || m == n)) {
-    svd_cusolver_gesvdjBatched(cloneBatchedColumnMajor(A), U, S, V, info, compute_uv);
+  const char* check_svd_doc = "Check doc at https://pytorch.org/docs/stable/generated/torch.linalg.svd.html";
+
+  const bool gesvdj_batched_requirements = (m <= 32 && n <= 32 && batch_size > 1 && (full_matrices || m == n));
+  bool convergence_check_needed = false;
+
+  if (!driver.has_value()) {
+    // use the default heuristics
+    if (gesvdj_batched_requirements) {
+      svd_cusolver_gesvdjBatched(cloneBatchedColumnMajor(A), U, S, V, info, compute_uv);
+      convergence_check_needed = true;
+    } else {
+      // gesvdj driver may be numerically unstable for large sized matrix
+      // Todo: Expose/adjust cusolver options of gesvdj number of iterations and numerical tolerance
+      svd_cusolver_gesvdj(cloneBatchedColumnMajor(A), U, S, V, info, full_matrices, compute_uv);
+      convergence_check_needed = true;
+    }
+    // DO NOT use gesvdaStridedBatched as a default SVD backend driver since it may be inaccurate.
   } else {
-    svd_cusolver_gesvdj(cloneBatchedColumnMajor(A), U, S, V, info, full_matrices, compute_uv);
+    if (driver.value() == "gesvd") {
+      svd_cusolver_gesvd(A, U, S, V, info, full_matrices, compute_uv);
+    } else if (driver.value() == "gesvdj") {
+      if (gesvdj_batched_requirements) {
+        svd_cusolver_gesvdjBatched(cloneBatchedColumnMajor(A), U, S, V, info, compute_uv);
+        convergence_check_needed = true;
+      } else {
+        // gesvdj driver may be numerically unstable for large sized matrix
+        // Todo: Expose/adjust cusolver options of gesvdj number of iterations and numerical tolerance
+        svd_cusolver_gesvdj(cloneBatchedColumnMajor(A), U, S, V, info, full_matrices, compute_uv);
+        convergence_check_needed = true;
+      }
+    } else if (driver.value() == "gesvda") {
+      // cuSOLVER: gesvdaStridedBatched is preferred for "tall skinny" matrices
+      TORCH_CHECK(batch_size > 0 && m > 0 && n > 0 && m > n,
+        "torch.linalg.svd: cuSOLVER `gesvda` driver requires the input matrices be non-empty "
+        "and number of rows be greater than number of columns (ideally tall and skinny).");
```

Comment:
Empty matrices are handled at the beginning of the `_linalg_svd`  function, so the first 3 checks should not be necessary.

About the `m > n`  condition, we should be able to work around it as we do for `gesvd` by computing `A^H`. See
We should be able to work around this, as we do in `gesvd`, by transposing (and perhaps taking conjs as well) the input matrix.
https://github.com/pytorch/pytorch/blob/bf730e5039e8bf1614de6ad08ac6a24a1622dcbe/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp#L438-L454

If this kernel does not support the `m == n` case (fair enough although a bit odd), this would make this torch check just report an error in that case.

Reply:
Thanks for the suggestions on empty input check!

I've once tried to transpose the input matrix in `gesvda`, like `gesvd` does, but it raised some cusolver errors. I'll try it again and either that works or I'll add some comments explaining this.